File: src/org/symbolmath/simplifier/Simplifier.java
package org.symbolmath.simplifier;

import com.sun.deploy.panel.ITreeNode;
import org.symbolmath.ast.expression.AdditiveExpression;
import org.symbolmath.parser.Parser;
import org.symbolmath.ast.*;
import org.symbolmath.scanner.Scanner;
import org.symbolmath.transformer.Transformer;

import java.util.List;

public class Simplifier {
  private static ExpressionSimplifier[] simplifiers = {
      new AdditiveExpressionSimplifier()
  };

  private ASTElement tree;

  public Simplifier(ASTElement tree) {
    this.tree = tree;
  }

  public Simplifier(String expression) {
    this(new Parser(expression, true).parse());
  }

  public ASTElement simplify() {
    SimplifyResult result;
    while ((result = simplify(tree)).simplified) {
      tree = result.newElement;
    }
    return tree;
  }

  private SimplifyResult simplify(ASTElement element) {
    boolean simplified = false;

    for (ExpressionSimplifier simplifier : simplifiers) {
      if (element.getClass().isAssignableFrom(simplifier.getApplicableElement())) {
        SimplifyResult result = simplifier.simplify(element);
        if (result.simplified) {
          simplified = true;
          if (result.newElement != null) {
            ASTElement parent = element.getParent();
            if (parent != null) {
              parent.replaceChild(element, result.newElement);
            } else {
              element = result.newElement;
            }
          }
        }
      }
    }

    if (!simplified) {
      List<? extends ASTElement> children = element.getChildren();
      for (ASTElement child : children) {
        SimplifyResult result = simplify(child);
        if (result.simplified) {
          simplified = true;
        }
      }
    }

    return new SimplifyResult(simplified, element);
  }

}

File: src/main/java/org/symbolmath/simplifier/Simplifier.kt
package org.symbolmath.simplifier

import org.symbolmath.ast.ASTElement
import org.symbolmath.parser.Parser

class Simplifier(private var tree: ASTElement) {
    constructor(expression: String) : this(Parser(expression, true).parse())

    fun simplify(): ASTElement {
        var result: SimplifyResult
        while (simplify(tree).also { result = it }.simplified) {
            tree = result.newElement
            println(tree)
        }
        return tree
    }

    private fun simplify(root: ASTElement): SimplifyResult {
        var element = root
        var simplified = false

        for (child in element.children) {
            if (simplify(child).simplified) {
                simplified = true
            }
        }

        val parent = element.parent

        for (simplifier in SIMPLIFIERS) {
            if (element.javaClass.isAssignableFrom(simplifier.getApplicableElement())) {
                val simplifiedElement = simplifier.simplify(element)
                if (!simplifiedElement.equals(element)) {
                    simplified = true
                    if (parent != null) {
                        parent.replaceChild(element, simplifiedElement)
                    } else {
                        element = simplifiedElement
                    }
                }
            }
        }

        return SimplifyResult(simplified, element)
    }

    companion object {
        private val SIMPLIFIERS = arrayOf(
            DivSimplifier() as ExpressionSimplifier<ASTElement>,
            MulCollapseSimplifier() as ExpressionSimplifier<ASTElement>,
            MulToPowSimplifier() as ExpressionSimplifier<ASTElement>,
            NegSimplifier() as ExpressionSimplifier<ASTElement>,
            AddCollapseSimplifier() as ExpressionSimplifier<ASTElement>,
            AddSimplifier() as ExpressionSimplifier<ASTElement>,
            AddZeroSimplifier() as ExpressionSimplifier<ASTElement>,
            PowerCollapseSimplifier() as ExpressionSimplifier<ASTElement>
        )
    }

    data class SimplifyResult(val simplified: Boolean, val newElement: ASTElement)
}